src/protobuf-net.AspNetCore/Formatters/ProtoInputFormatter.cs

```diff
@@ -0,0 +1,109 @@
+        public ProtoInputFormatter(TypeModel model = null)
+        {
+            _model = model ?? RuntimeTypeModel.Default;
+        }
+
+        /// <inheritdoc/>
+        protected override bool CanReadType(Type type)
+            => _model.CanSerializeContractType(type);
+
+        /// <inheritdoc/>
+        public override Task<InputFormatterResult> ReadAsync(InputFormatterContext context)
+        {
+            // default implementation does length checks; we *really* don't want those, since
+            // zero length is a perfectly valid payload in protobuf
+            return ReadRequestBodyAsync(context);
+        }
+
+        /// <inheritdoc/>
+        public override Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterContext context)
+        {
+            // note: *today*, protobuf-net lacks fully async read, so this buffers in the pipe until
+            // we get the Content-Length that was advertised by the caller
+
+            if (context is null) throw new ArgumentNullException(nameof(context));
+            var request = context.HttpContext.Request;
+            var length = request.ContentLength ?? -1;
+            if (length < 0) throw new InvalidOperationException("Invalid or missing content-length");
+            var reader = request.BodyReader;
+            var type = context.ModelType;
+
+            // try and read synchronously
+            if (reader.TryRead(out var readResult) && ProcessReadBuffer(type, reader, length, readResult, out var payload))
+            {
+                return InputFormatterResult.SuccessAsync(payload);
+            }
+            return ReadRequestBodyAsyncSlow(type, reader, length);
+        }
+
+        private bool ProcessReadBuffer(Type type, PipeReader reader, long length, in ReadResult readResult, out object payload)
+        {
+            if (readResult.IsCanceled) throw new OperationCanceledException();
+            var buffer = readResult.Buffer;
+            if (ParseIfComplete(type, ref buffer, length, out payload))
+            {
+                // mark consumed
+                reader.AdvanceTo(buffer.End, buffer.End);
+                return true;
+            }
+            else if (readResult.IsCompleted)
+            {
+                throw new InvalidOperationException($"Incomplete protobuf payload received; got {buffer.Length} of {length} bytes");
+            }
+            // not enough data; put it back by saying we looked at everything and took nothing
+            reader.AdvanceTo(buffer.Start, buffer.End);
+            return false;
+        }
+
+        private async Task<InputFormatterResult> ReadRequestBodyAsyncSlow(Type type, PipeReader reader, long length)
+        {
+            while (true)
+            {
+                if (!reader.TryRead(out var readResult))
+                {
+                    readResult = await reader.ReadAsync().ConfigureAwait(false);
```

Comment:
ConfigureAwait not needed

Reply:
Because you have no sync-context? K, I can kill that. Force of habit from a library author who doesn't usually know the run context.